# Ex1 - Filtering and Sorting Data

This time we are going to pull data directly from the internet.
Special thanks to: https://github.com/justmarkham for sharing the dataset and materials.

### Step 1. Import the necessary libraries

In [53]:
import pandas as pd

### Step 2. Import the dataset from this [address](https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv). 

In [54]:
file_path = r'https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv'

### Step 3. Assign it to a variable called chipo.

In [55]:
chipo = pd.read_csv('https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv', sep='\t')

In [56]:
chipo.head(10)

order_id  quantity                              item_name  \
0         1         1           Chips and Fresh Tomato Salsa   
1         1         1                                   Izze   
2         1         1                       Nantucket Nectar   
3         1         1  Chips and Tomatillo-Green Chili Salsa   
4         2         2                           Chicken Bowl   
5         3         1                           Chicken Bowl   
6         3         1                          Side of Chips   
7         4         1                          Steak Burrito   
8         4         1                       Steak Soft Tacos   
9         5         1                          Steak Burrito   

                                  choice_description item_price  
0                                                NaN     $2.39   
1                                       [Clementine]     $3.39   
2                                            [Apple]     $3.39   
3                                                NaN     $2.39   
4  [Tomatillo-Red Chili Salsa (Hot), [Black Beans...    $16.98   
5  [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...    $10.98   
6                                                NaN     $1.69   
7  [Tomatillo Red Chili Salsa, [Fajita Vegetables...    $11.75   
8  [Tomatillo Green Chili Salsa, [Pinto Beans, Ch...     $9.25   
9  [Fresh Tomato Salsa, [Rice, Black Beans, Pinto...     $9.25

### Step 4. How many products cost more than $10.00?

In order to be able to use comparison operators, we need to make sure the column item_price is of a numerical type.The item_column contains values in a string format using a dollar sign. We need to clean the data by removing the dollar sign and turn the string values to float.
1. Remove the Dollar Sign and Convert to Float:

In [57]:
chipo['item_price'] = chipo["item_price"].str.replace("$", "").astype(float)

Checking if there are missing values in the column : 

In [58]:
# Check for missing values in 'item_price' column
missing_values = chipo["item_price"].isnull() 

In [59]:
missing_values

0       False
1       False
2       False
3       False
4       False
        ...  
4617    False
4618    False
4619    False
4620    False
4621    False
Name: item_price, Length: 4622, dtype: bool

In [60]:
chipo.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   order_id            4622 non-null   int64  
 1   quantity            4622 non-null   int64  
 2   item_name           4622 non-null   object 
 3   choice_description  3376 non-null   object 
 4   item_price          4622 non-null   float64
dtypes: float64(1), int64(2), object(2)
memory usage: 180.7+ KB


Length is 4622 and entries 4622, looks like there no missing Nan Values.

### Notes

- Step 2: Handle any potential missing values in item_price (if any).For example, we could fill missing values with 0 or drop them
- chipo['item_price'] = chipo['item_price'].fillna(0)  # if we wanted to fill with 0
- chipo = chipo.dropna(subset=['item_price'])  # if we wanted to drop rows with missing item_price


##### Now we can filter items that cost more than $10.00

Checking if we have successfuly removed $ sign and we are able to use comparison operator now  :

In [61]:
filtered_items = chipo[chipo['item_price'] > 10.00]

In [62]:
filtered_items

order_id  quantity           item_name  \
4            2         2        Chicken Bowl   
5            3         1        Chicken Bowl   
7            4         1       Steak Burrito   
13           7         1        Chicken Bowl   
23          12         1     Chicken Burrito   
...        ...       ...                 ...   
4610      1830         1       Steak Burrito   
4611      1830         1      Veggie Burrito   
4617      1833         1       Steak Burrito   
4618      1833         1       Steak Burrito   
4619      1834         1  Chicken Salad Bowl   

                                     choice_description  item_price  
4     [Tomatillo-Red Chili Salsa (Hot), [Black Beans...       16.98  
5     [Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...       10.98  
7     [Tomatillo Red Chili Salsa, [Fajita Vegetables...       11.75  
13    [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
23    [[Tomatillo-Green Chili Salsa (Medium), Tomati...       10.98  
...                                                 ...         ...  
4610  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4611  [Tomatillo Green Chili Salsa, [Rice, Fajita Ve...       11.25  
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour ...       11.75  
4618  [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...       11.75  
4619  [Fresh Tomato Salsa, [Fajita Vegetables, Pinto...       11.25  

[1130 rows x 5 columns]

#### Notes :
- df['item_price'].str.replace('$', ''): This removes the dollar sign from each value in the item_price column. The str.replace() function is used perform string replacement.
- .astype(float): Converts the resulting string values to float. This allows for numerical comparisons.

### Step 5. What is the price of each item? 
###### print a data frame with only two columns item_name and item_price

In [63]:
print(chipo[['item_name', 'item_price']])

                                  item_name  item_price
0              Chips and Fresh Tomato Salsa        2.39
1                                      Izze        3.39
2                          Nantucket Nectar        3.39
3     Chips and Tomatillo-Green Chili Salsa        2.39
4                              Chicken Bowl       16.98
...                                     ...         ...
4617                          Steak Burrito       11.75
4618                          Steak Burrito       11.75
4619                     Chicken Salad Bowl       11.25
4620                     Chicken Salad Bowl        8.75
4621                     Chicken Salad Bowl        8.75

[4622 rows x 2 columns]


### Step 6. Sort by the name of the item

In [64]:
chipo.sort_values("item_name",ascending = True, inplace = False)

order_id  quantity          item_name  \
3389      1360         2  6 Pack Soft Drink   
341        148         1  6 Pack Soft Drink   
1849       749         1  6 Pack Soft Drink   
1860       754         1  6 Pack Soft Drink   
2713      1076         1  6 Pack Soft Drink   
...        ...       ...                ...   
2384       948         1  Veggie Soft Tacos   
781        322         1  Veggie Soft Tacos   
2851      1132         1  Veggie Soft Tacos   
1699       688         1  Veggie Soft Tacos   
1395       567         1  Veggie Soft Tacos   

                                     choice_description  item_price  
3389                                        [Diet Coke]       12.98  
341                                         [Diet Coke]        6.49  
1849                                             [Coke]        6.49  
1860                                        [Diet Coke]        6.49  
2713                                             [Coke]        6.49  
...                                                 ...         ...  
2384  [Roasted Chili Corn Salsa, [Fajita Vegetables,...        8.75  
781   [Fresh Tomato Salsa, [Black Beans, Cheese, Sou...        8.75  
2851  [Roasted Chili Corn Salsa (Medium), [Black Bea...        8.49  
1699  [Fresh Tomato Salsa, [Fajita Vegetables, Rice,...       11.25  
1395  [Fresh Tomato Salsa (Mild), [Pinto Beans, Rice...        8.49  

[4622 rows x 5 columns]

### Step 7. What was the quantity of the most expensive item ordered?

We can first find the index of the most expensive item :

In [65]:
max_price_index = chipo['item_price'].idxmax()

In [66]:
# Use the index to get the corresponding row
most_expensive_item = chipo.loc[max_price_index]


In [67]:
most_expensive_item

order_id                                      1443
quantity                                        15
item_name             Chips and Fresh Tomato Salsa
choice_description                             NaN
item_price                                   44.25
Name: 3598, dtype: object

### Step 8. How many times was a Veggie Salad Bowl ordered?

#### Use groupby() method :

We can use groupby method to create an object where the data is grouped by each unique value in the column "item_name".The groupby method splits the DataFrame into subsets (groups) based on the unique values in the item_name column. 

- ['quantity']: With this we narrow down the operation to only the "quantity" group within each group.

- sum(): This returns a new pandas Series where the index is the unique values from item_name and the values are the sum of quantity for each group.

In [68]:
item_order_counts = chipo.groupby('item_name')['quantity'].sum()

#### Use .loc[] to get the count for the specific item.

In [76]:
item_order_counts.loc['Veggie Salad Bowl']

18

### An alternative approach :
#### Is it possible to find how many times the Veggie Salad Bowl was ordered without grouping and summing in seperate steps?  

In [74]:
veggie_salad_bowl_orders = chipo.loc[chipo['item_name'] == 'Veggie Salad Bowl', 'quantity'].sum()

1. First , we filter based on condition : the [chipo['item_name] == "Veggie Salad Bowl] : This creates a boolean mask that is True for rows with item_name 'Veggie Salad Bowl' , and False for rows otherwise.
2. chipo.loc[chipo['item_name'] == 'Veggie Salad Bowl', 'quantity'] ---> Using .loc[] we filter the Dataframe Chipo to include only rows where the item_name is Veggie Salad Bowl .From these filtered rows we select the "quantity" column.
3. sum() : .sum(): This sums up all the quantity values that match the condition, which gives us the total number of times 'Veggie Salad Bowl' was ordered.

In [75]:
print(f"The Veggie Salad Bowl was ordered {veggie_salad_bowl_orders} times")

The Veggie Salad Bowl was ordered 18 times


### Step 9. How many times did someone order more than one Canned Soda?

1. Filter the rows that include only items_name == Canned Soda

In [78]:
soda_orders = chipo[chipo['item_name'] == "Canned Soda"]

In [79]:
soda_orders

order_id  quantity    item_name choice_description  item_price
18           9         2  Canned Soda           [Sprite]        2.18
28          14         1  Canned Soda       [Dr. Pepper]        1.09
51          23         2  Canned Soda     [Mountain Dew]        2.18
53          24         1  Canned Soda           [Sprite]        1.09
107         47         1  Canned Soda       [Dr. Pepper]        1.09
...        ...       ...          ...                ...         ...
3897      1562         1  Canned Soda     [Mountain Dew]        1.09
3926      1575         1  Canned Soda       [Dr. Pepper]        1.09
3936      1578         1  Canned Soda  [Diet Dr. Pepper]        1.09
4008      1604         1  Canned Soda        [Diet Coke]        1.09
4051      1621         1  Canned Soda           [Sprite]        1.09

[104 rows x 5 columns]

##### Once we have soda_orders, we can perform additional operations like summing the quantities (soda_orders['quantity'].sum()) or finding rows where the quantity exceeds a certain threshold (soda_orders[soda_orders['quantity'] > 1]).

Then we need to filter quantity > 1 :

In [84]:
excess_amount_soda_orders = soda_orders[soda_orders['quantity'] > 1]

In [87]:
print("Orders with more than 1 Sodas':")
print(f"{excess_amount_soda_orders}")

Orders with more than 1 Sodas':
      order_id  quantity    item_name choice_description  item_price
18           9         2  Canned Soda           [Sprite]        2.18
51          23         2  Canned Soda     [Mountain Dew]        2.18
162         73         2  Canned Soda        [Diet Coke]        2.18
171         76         2  Canned Soda  [Diet Dr. Pepper]        2.18
350        150         2  Canned Soda        [Diet Coke]        2.18
352        151         2  Canned Soda        [Coca Cola]        2.18
698        287         2  Canned Soda        [Coca Cola]        2.18
700        288         2  Canned Soda        [Coca Cola]        2.18
909        376         2  Canned Soda     [Mountain Dew]        2.18
1091       450         2  Canned Soda       [Dr. Pepper]        2.18
1092       450         2  Canned Soda        [Coca Cola]        2.18
1944       787         2  Canned Soda       [Dr. Pepper]        2.18
2135       859         2  Canned Soda        [Diet Coke]        2.18
22

We now have all  the orders with more than one Sodas. But how many of these are?

In [88]:
print("The number of orders with more than one Sodas are: ", len(excess_amount_soda_orders))

The number of orders with more than one Sodas are:  20
